src/pages/tutorial/cw-contract/query.mdx

```diff
@@ -0,0 +1,323 @@
+but there are many different entities in a single file, and I think we can do better. I can already
+distinguish three different types of entities in the code: entry points, messages, and handlers. In
+most contracts, we would divide them across three files. Start with extracting all the messages to
+`src/msg.rs`:
+
+```rust copy filename="src/msg.rs" template="empty"
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize)]
+pub struct GreetResp {
+    pub message: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub enum QueryMsg {
+    Greet {},
+}
+```
+
+You probably noticed that I made my `GreetResp` fields public. It is because they have to be now
+accessed from a different module. Now move forward to the `src/contract.rs` file:
+
+```rust {1} copy filename="src/contract.rs"
+use crate::msg::{GreetResp, QueryMsg};
+use cosmwasm_std::{
+    to_json_binary, Binary, Deps, DepsMut, Empty, Env, MessageInfo, Response, StdResult,
+};
+
+pub fn instantiate(
+    _deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    _msg: Empty,
+) -> StdResult<Response> {
+    Ok(Response::new())
+}
+
+pub fn query(_deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    use QueryMsg::*;
+
+    match msg {
+        Greet {} => to_json_binary(&query::greet()?),
+    }
+}
+
+mod query {
+    use super::*;
+
+    pub fn greet() -> StdResult<GreetResp> {
+        let resp = GreetResp {
+            message: "Hello World".to_owned(),
+        };
+
+        Ok(resp)
+    }
+}
+```
+
+I moved most of the logic here, so my `src/lib.rs` is just a very thin library entry with nothing
+else but module definitions and entry points definition. I removed the `#[entry_point]` attribute
+from my `query` function in `src/contract.rs`. I will have a function with this attribute. Still, I
+wanted to split functions' responsibility further - not the `contract::query` function is the
+top-level query handler responsible for dispatching the query message. The `query` function on
+crate-level is only an entry point. It is a subtle distinction, but it will make sense in the future
+when we would like not to generate the entry points but to keep the dispatching functions. I
+introduced the split now to show you the typical contract structure.
+
+Now the last part, the `src/lib.rs` file:
+
+```rust copy filename="src/lib.rs"
+use cosmwasm_std::{
+    entry_point, Binary, Deps, DepsMut, Empty, Env, MessageInfo, Response, StdResult,
+};
+
+mod contract;
+mod msg;
+
+#[entry_point]
+pub fn instantiate(deps: DepsMut, env: Env, info: MessageInfo, msg: Empty)
+  -> StdResult<Response>
+{
+    contract::instantiate(deps, env, info, msg)
+}
+
+#[entry_point]
+pub fn query(deps: Deps, env: Env, msg: msg::QueryMsg)
+  -> StdResult<Binary>
+{
+    contract::query(deps, env, msg)
+}
+```
+
+Straightforward top-level module. Definition of submodules and entry points, nothing more.
+
+Now, when we have the contract ready to do something, let's go and test it.
```

Comment:
```suggestion
Now, since our contract is ready to do something, let's go and test it.
```

I'm guessing this better conveys what you meant, could be wrong